src/mono/mono/mini/mono-private-unstable.h

```diff
@@ -47,6 +67,10 @@ monovm_initialize (int propertyCount, const char **propertyKeys, const char **pr
 MONO_API int 
 monovm_runtimeconfig_initialize (MonovmRuntimeConfigArguments *arg, MonovmRuntimeConfigArgumentsCleanup cleanup_fn, void *user_data);
 
+// The wrapper MonoCoreRuntimeProperties struct can be stack-allocated or freed, but any of the structs inside it will *not* be copied and need to be heap-allocated
+MONO_API int
+monovm_initialize_preparsed (MonoCoreRuntimeProperties *parsed_properties, int propertyCount, const char **propertyKeys, const char **propertyValues);
```

Comment:
we can try to futureproof this by asking for a `sizeof (MonoCoreRuntimeProperties)` argument (assuming that in the future we only ever add fields and never remove them)

Reply:
I mean I understand this is an unstable header, but this will let us push runtime changes without having a Xamarin.Android PR ready as soon as they take a new preview of the runtime.

Reply:
I had that originally, but I decided we want to just not allow people to link against a different version of the runtime than they built against. If you do that, you're on your own. Unless I'm missing a reason that you'd want that?

Reply:
They wouldn't need to update the code as long as they're building against the new runtime though? Unless they're declaring the struct separately instead of using our header, which we also really don't want.

Reply:
The struct could have its first field to specify its own size - similar to how sockaddr works in POSIX. The caller would have to set that property and MonoVM would be able to tell the struct "version" this way. Nothing should be ever removed or rearranged, things could be added only at the end of the struct. I think it would make for nice future proofing, fully agreed with @lambdageek

Reply:
I'm confused - what scenario is the size supposed to solve here? What are we future proofing against?

As far as I know X-A no longer redeclares all the functions it uses and can just use our headers directly, and we make no ABI guarantees so you *must* build and link against the same version of Mono. Given that, your build isn't broken as long as we don't remove elements from the struct, even if we rearrange them, so what would we be future-proofing against?

I'm happy to add this sort of thing if there's a clear scenario where it would add value, but I don't see one in a net6 world.

Reply:
How did/does the XA Shared Runtime stuff work? Does it rely on an ABI guarantee or is there some versioning built in?

Reply:
> How did/does the XA Shared Runtime stuff work? Does it rely on an ABI guarantee or is there some versioning built in?
> 
> Otherwise it does seem like there's not much need for this.

You mean the shared runtime we install on debug builds? We dropped that. As for XA proper, we link directly against `libmonosgen-2.0.so`

Reply:
> I'm confused - what scenario is the size supposed to solve here? What are we future proofing against?
> 
> As far as I know X-A no longer redeclares all the functions it uses and can just use our headers directly, and we make no ABI guarantees so you _must_ build and link against the same version of Mono. Given that, your build isn't broken as long as we don't remove elements from the struct, even if we rearrange them, so what would we be future-proofing against?
> 
> I'm happy to add this sort of thing if there's a clear scenario where it would add value, but I don't see one in a net6 world.

My only worry is what happens if someone locally overrides the NET6 Mono Android runtime nugets so that Xamarin.Android applications will suddenly be **compiled** against a different version of the MonoVM runtime and running against **another** MonoVM version. In the "legacy" world we always shipped Mono with XA, so the chance for above was slim to none. In the nuget world... you never know.

Reply:
I think I've convinced myself that I'm not too worried about that scenario, and if it comes up we should have a more proper, global solution rather than trying to patch functions like this. At the end of the day, we aren't making ABI guarantees and if people are violating that, it's on them.